compositor_pipeline/src/queue/utils.rs

```diff
@@ -0,0 +1,90 @@
+use std::{
+    collections::VecDeque,
+    mem,
+    time::{Duration, Instant},
+};
+
+use compositor_render::{AudioSamplesBatch, Frame};
+
+use super::DEFAULT_BUFFER_DURATION;
+
+#[derive(Debug)]
+pub(super) enum InputState<Payload: ApplyOffsetExt> {
+    WaitingForStart,
+    Buffering { buffer: Vec<(Payload, Duration)> },
+    Ready { offset: chrono::Duration },
+}
+
+impl<Payload: ApplyOffsetExt> InputState<Payload> {
+    pub(super) fn process_new_chunk(
+        &mut self,
+        mut payload: Payload,
+        pts: Duration,
+        clock_start: Instant,
+    ) -> VecDeque<Payload> {
```

Comment:
1. This can just return iterator
2. If the input is shorter than `DEFAULT_BUFFER_DURATION` it won't work. If the stream freezes after receiving only a few frames, the queue won't enqueue them.
3. This + not accounting for buffer duration in the video queue thread results in black frames for the beginning of the input stream.

Reply:
1. Yes, but it still needs to be an iterator that owns the data, so what would be the point. Plus this way I can just call append on queue
2. Good point, but I don't think this is a problem.
3. It does. Please explain why you think that and I will maybe add a description to clarify that.

Reply:
1.
I don't like that the typing of this function depends on the type in which the queue keeps frames/samples.

I thought before about keeping the audio samples in some structure that is ordered based on their pts (like ordered map / BST tree), because the audio decoder doesn't guarantee monotonical timestamps (RTP packets can be received non-monotonically and audio PTS come directly from as RTP packets PTSs).
Using the concrete `VecDeque` type here makes this code less flexible (sth like using `&Vec<T>` instead of `&[T]`)

In any case like this using iter and extend is more flexible, but if this iter would need to own data it comes with some overhead. 

I understand why you might want to keep this instead, so I leave it up to you.

2. If we don't want to use larger buffers in the near future
3. Discussed offline